ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java

```diff
@@ -336,64 +339,76 @@ private static Object extractRowKeyWhereClause(
-    final Instant lowerBound = extractWindowBound(lower);
-    return Range.closed(lowerBound, upperBound);
+    return extractWindowBound(lower, upper);
   }
 
-  private static Type getBoundType(final ComparisonExpression comparison) {
+  private static Type getSimplifiedBoundType(final ComparisonExpression comparison) {
     final Type type = comparison.getType();
     final boolean inverted = comparison.getRight() instanceof ColumnReferenceExp;
 
     switch (type) {
       case LESS_THAN:
-        return inverted ? Type.GREATER_THAN : type;
       case LESS_THAN_OR_EQUAL:
-        return inverted ? Type.GREATER_THAN_OR_EQUAL : type;
+        return inverted ? Type.GREATER_THAN : Type.LESS_THAN;
       case GREATER_THAN:
-        return inverted ? Type.LESS_THAN : type;
       case GREATER_THAN_OR_EQUAL:
-        return inverted ? Type.LESS_THAN_OR_EQUAL : type;
+        return inverted ? Type.LESS_THAN : Type.GREATER_THAN;
       default:
         return type;
     }
   }
 
-  private static Instant extractWindowBound(final ComparisonExpression comparison) {
-    final Expression other = comparison.getRight() instanceof ColumnReferenceExp
-        ? comparison.getLeft()
-        : comparison.getRight();
+  private static Range<Instant> extractWindowBound(
+      final Optional<ComparisonExpression> lowerComparison,
+      final Optional<ComparisonExpression> upperComparison
+  ) {
+    if (!lowerComparison.isPresent() && !upperComparison.isPresent()) {
+      return Range.all();
+    }
 
-    final Instant bound = asInstant(other);
+    if (!lowerComparison.isPresent()) {
+      final Instant upper = asInstant(getNonColumnRefSide(upperComparison.get()));
+      final BoundType upperType = getRangeBoundType(upperComparison.get());
+      return Range.upTo(upper, upperType);
+    }
 
-    switch (getBoundType(comparison)) {
-      case LESS_THAN:
-        return bound.minusMillis(1);
-      case GREATER_THAN:
-        return bound.plusMillis(1);
-      default:
-        return bound;
+    if (!upperComparison.isPresent()) {
+      final Instant lower = asInstant(getNonColumnRefSide(lowerComparison.get()));
+      final BoundType lowerType = getRangeBoundType(lowerComparison.get());
+      return Range.downTo(lower, lowerType);
     }
+
+    final Instant lower = asInstant(getNonColumnRefSide(lowerComparison.get()));
+    final BoundType lowerType = getRangeBoundType(lowerComparison.get());
+
+    final Instant upper = asInstant(getNonColumnRefSide(upperComparison.get()));
+    final BoundType upperType = getRangeBoundType(upperComparison.get());
+
+    return Range.range(lower, lowerType, upper, upperType);
+  }
+
+  private static BoundType getRangeBoundType(final ComparisonExpression lowerComparison) {
+    final boolean openBound = lowerComparison.getType() == Type.LESS_THAN
+        || lowerComparison.getType() == Type.GREATER_THAN;
+
+    return openBound
+        ? BoundType.OPEN
```

Comment:
don't we only have open bounds? Because of `getSimplifiedBoundType`.

Reply:
Never mind - I see that's just grouping the open and closed comparisons of the same type together under a common key. The type is still preserved though.